### PR TITLE
Uptime

### PR DESCRIPTION
**Pull request by timzimmermann**
Add a table that shows the time passed since the last boot.

See #1273.

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by marpaia** [comment]
Can you move all of this into the same c++ file and use ifdef's to switch between the minimum set of OS X functionality vs linux functionality.

Also, you can/should merge `long osquery::getUptime` and `double osquery::getUptime` into a single symbol.

File: osquery/tables/system/uptime_utils.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  Row r;
+  r["days"] = INTEGER(total_seconds / 60 / 60 / 24);
+  r["hours"] = INTEGER((total_seconds / 60 / 60) % 24);
+  r["minutes"] = INTEGER((total_seconds / 60) % 60);
+  r["seconds"] = INTEGER(total_seconds % 60);
+  r["total_seconds"] = INTEGER(total_seconds);
+  return r;
+}
+
+QueryData genUptimeRow(long uptime_in_seconds) {
```
Comment (marpaia) [inline osquery/tables/system/uptime_utils.cpp]: this won't be necessary if all of the platform specific switching occurs in `getUptime`

File: osquery/tables/system/uptime_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#include <osquery/tables.h>
+
+#include "osquery/tables/system/uptime_utils.h"
+
+namespace osquery {
+namespace tables {
+
+Row secondsToTimeRow(long total_seconds) {
```
Comment (marpaia) [inline osquery/tables/system/uptime_utils.cpp]: same here

**Comment by marpaia** [comment]
Also, please add the following to the top of each file, or else lawyers will burn us all at the stake :)

```
/*
 *  Copyright (c) 2014, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */
```

**Comment by timzimmermann** [comment]
Sure, check 020c0d9.
I just thought it was cleaner to split it.

**Comment by theopolis** [comment]
Random thought, would a column "uptime", that included the update as a count of seconds, added to the time table be sufficient?

**Comment by timzimmermann** [comment]
Well, it depends on how you want to use it, right?
You could then argue that the time table could also just contain the unix timestamp in one column.
I prefer the multiple columns because it is much more readable for humans. However, if more people prefer a one-column solution, we can go with that.

In addition, I was thinking about extending the time table to also show date information, but I will put that into a separate issue later.

**Comment by theopolis** [comment]
Certainly! If you queried: "select minutes  from uptime" would you expect total, or relative to the hour? I'm not sure a modulus, or relative, number is intuitive. Unless! You select * and observe the results, then it becomes obvious. I really like uptime as something we can join against at query time, and don't mind having to specify "total_seconds" as the column. But it does seem a to add a bit of friction compared to a "select uptime from time".

I'm up in the air, your call! :)

By the way, that last commit is looking good. If you can squash the set of commits in this PR into 1, and you're happy with the implementation, we can merge!

File: osquery/tables/system/uptime.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ */
+
+#include <osquery/tables.h>
+
+#if defined(__APPLE__)
+  #include <time.h>
+  #include <errno.h>
+  #include <sys/sysctl.h>
+#elif defined(__linux) || defined(__linux__) || defined(linux)
```
Comment (marpaia) [inline osquery/tables/system/uptime.cpp]: @theopolis IIRC, don't we define a LINUX macro or something somewhere for times like this?

Comment (marpaia) [inline osquery/tables/system/uptime.cpp]: yeah, all other places we just do `#ifdef __linux__`, i would opt to stay consistent with that, seems to be working.

Comment (theopolis) [inline osquery/tables/system/uptime.cpp]: We do have an undocumented set of defines that allow us to specify the platform from CMake logic. One day we should codemod out all the compiled-based decision logic (e.g.: `gcc -dM -E - < /dev/null`) and prefer our CMake decisions as the source of truth.

**Comment by timzimmermann** [comment]
I took some time to think about your point, @theopolis.
I agree that there might be some confusion. However, I decided that it is still the way to go, for two reasons:
 * I don't feel that an uptime column in the time table is very intuitive. I myself wouldn't expect it there.
 * With regards to the additional information that is available, I think that it brings more of a benefit than it might cause confusion.

I also squashed the commits into one. From my side it is ready to merge. Thank you for your help!

**Comment by theopolis** [comment]
ok to test

**Comment by theopolis** [comment]
Thanks @timzimmermann, very good points. This is looking excellent, if Jenkins agrees we're ready to rock!

**Comment by marpaia** [comment]
Woot, thanks @timzimmermann!

**Pull request merged**